crates/factors/src/runtime_factors.rs

```diff
@@ -81,6 +81,12 @@ pub trait RuntimeFactors: Sized + 'static {
     ) -> Option<Option<&mut F::InstanceBuilder>>;
 }
 
+/// Allows querying an `InstanceBuilders` for a particular `Factor`'s `InstanceBuilder`.
+pub trait HasInstanceBuilder {
```

Comment:
As in treating this as the trait all `InstanceBuilders` implement? I guess that could work.

Reply:
Are we fine with getting rid of the `InstanceBuilders` struct? It seems the role of that struct is to play the same role as this trait.

Reply:
That struct is slightly different in that it represents InstanceBuilders in the middle of being initialized, so it can return `Error::DependencyOrderingError`. If `for_factor` returned a `Result` then replacing it would make sense.

Reply:
Ok - it would make life much easier if we renamed the struct (otherwise we'll have to deal with a lot of situations where the struct and the trait would conflict. Do you have an idea for a different name for the struct?

Reply:
Its only used as an argument to `prepare`, so...`PrepareInstanceBuilders`? I would have just made it a field on `PrepareContext` but mixing the `&mut` in there made juggling refs very awkward in practice.

Reply:
>  I would have just made it a field on `PrepareContext` but mixing the `&mut` in there made juggling ref ownership very awkward in practice.

You know after pondering this a bit more I think I was actually just doing this wrong when I tried it before. I had e.g.

```rust
impl<'a F: Factor> PrepareContext<'a, F> {
  pub fn app_state(&self) -> &F::AppState { ... }
}
```

but I could/should have made it `pub fn app_state(&self) -> &'a F::AppState`. Since immutable refs are `Copy`, that wouldn't hold the borrow on `&self`...

All of which is to say that I think `InstanceBuilders` probably _can_ be merged into `PrepareContext` after all.

Reply:
Yeah that seems to work fine: https://github.com/fermyon/spin/pull/2792

Reply:
Rename this now?